Clang fails to parse mp_units

Hi,

since e9972debc98ce5d00db47409248bbcf06fafaf73 (#161671) clang isn't able to parse mp_units anymore as shown here: https://godbolt.org/z/4seG47hYK
<details>

```
In file included from <source>:1:
In file included from /opt/compiler-explorer/libs/mp-units/v2.4.0/src/systems/include/mp-units/systems/si/units.h:30:
In file included from /opt/compiler-explorer/libs/mp-units/v2.4.0/src/core/include/mp-units/framework/construction_helpers.h:26:
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/core/include/mp-units/framework/quantity_point.h:144:25: error: deduced type 'struct absolute_zero' does not satisfy 'PointOriginFor<get_quantity_spec(mp_units::si::milli_<mp_units::si::kelvin>{})>'
  144 | [[nodiscard]] consteval PointOriginFor<get_quantity_spec(R{})> auto default_point_origin(R)
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/core/include/mp-units/framework/construction_helpers.h:52:79: note: in instantiation of function template specialization 'mp_units::default_point_origin<mp_units::si::milli_<mp_units::si::kelvin>>' requested here
   52 |   [[nodiscard]] constexpr quantity_point<MP_UNITS_EXPRESSION_WORKAROUND(R{}), default_point_origin(R{}), Rep>
      |                                                                               ^
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/core/include/mp-units/framework/construction_helpers.h:65:57: note: in instantiation of template class 'mp_units::absolute_<mp_units::si::milli_<mp_units::si::kelvin>>' requested here
   65 | constexpr absolute_<MP_UNITS_REMOVE_CONST(decltype(R))> absolute{};
      |                                                         ^
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/systems/include/mp-units/systems/si/units.h:81:65: note: in instantiation of variable template specialization 'mp_units::absolute' requested here
   81 | inline constexpr struct ice_point final : relative_point_origin<absolute<milli<kelvin>>(273'150)> {} ice_point;
      |                                                                 ^
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/core/include/mp-units/framework/quantity_point_concepts.h:76:44: note: because 'QuantitySpecOf<decltype(kind_of_<decltype(struct thermodynamic_temperature{{{{}}}})>{{{{}}}}), struct absolute_zero::_quantity_spec_>' evaluated to false
   76 | concept PointOriginFor = PointOrigin<T> && QuantitySpecOf<MP_UNITS_REMOVE_CONST(decltype(QS)), T::_quantity_spec_>;
      |                                            ^
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/core/include/mp-units/framework/quantity_spec_concepts.h:141:22: note: because 'decltype(struct thermodynamic_temperature{{{{}}}})' (aka 'struct thermodynamic_temperature') does not satisfy 'QuantitySpec'
  141 |   QuantitySpec<T> && QuantitySpec<MP_UNITS_REMOVE_CONST(decltype(QS))> && detail::QuantitySpecConvertibleTo<T{}, QS> &&
      |                      ^
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/core/include/mp-units/framework/quantity_spec_concepts.h:40:24: note: because 'mp_units::isq::thermodynamic_temperature &' does not satisfy 'SymbolicConstant'
   40 | concept QuantitySpec = detail::SymbolicConstant<T> && std::derived_from<T, detail::quantity_spec_interface_base>;
      |                        ^
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/core/include/mp-units/ext/type_traits.h:199:28: note: because 'std::is_empty_v<mp_units::isq::thermodynamic_temperature &>' evaluated to false
  199 | concept SymbolicConstant = std::is_empty_v<T> && std::is_final_v<T>;
      |                            ^
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/core/include/mp-units/ext/type_traits.h:199:28: note: 'mp_units::isq::thermodynamic_temperature &' is not empty
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/core/include/mp-units/ext/type_traits.h:199:28: note: because it is a reference type
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/systems/include/mp-units/systems/isq/base_quantities.h:53:15: note: 'thermodynamic_temperature' defined here
   53 | QUANTITY_SPEC(thermodynamic_temperature, dim_thermodynamic_temperature);
      |               ^
In file included from <source>:1:
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/systems/include/mp-units/systems/si/units.h:81:65: error: type 'const absolute_<decltype(milli_<decltype(struct kelvin{{{}}})>{{{{}, {}}}})>' (aka 'const absolute_<milli_<struct kelvin>>') does not provide a call operator
   81 | inline constexpr struct ice_point final : relative_point_origin<absolute<milli<kelvin>>(273'150)> {} ice_point;
      |                                                                 ^~~~~~~~~~~~~~~~~~~~~~~
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/systems/include/mp-units/systems/si/units.h:81:102: error: constexpr variable 'ice_point' must be initialized by a constant expression
   81 | inline constexpr struct ice_point final : relative_point_origin<absolute<milli<kelvin>>(273'150)> {} ice_point;
      |                                                                                                      ^~~~~~~~~
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/systems/include/mp-units/systems/si/units.h:82:23: error: constexpr variable 'zeroth_degree_Celsius' must be initialized by a constant expression
   82 | inline constexpr auto zeroth_degree_Celsius = ice_point;
      |                       ^                       ~~~~~~~~~
/opt/compiler-explorer/libs/mp-units/v2.4.0/src/systems/include/mp-units/systems/si/units.h:83:95: error: non-type template argument is not a constant expression
   83 | inline constexpr struct degree_Celsius final : named_unit<symbol_text{u8"℃", "`C"}, kelvin, zeroth_degree_Celsius> {} degree_Celsius;
      |                                                                                             ^~~~~~~~~~~~~~~~~~~~~
5 errors generated.
Compiler returned: 1
```
</details>

CC @cor3ntin

--- a/clang/lib/Sema/TreeTransform.h
+++ b/clang/lib/Sema/TreeTransform.h
@@ -4297,6 +4297,15 @@ public:
     return getSema().OpenACC().ActOnOpenACCAsteriskSizeExpr(AsteriskLoc);
   }
 
+  ExprResult
+  RebuildSubstNonTypeTemplateParmExpr(Decl *AssociatedDecl,
+                                      const NonTypeTemplateParmDecl *NTTP,
+                                      SourceLocation Loc, TemplateArgument Arg,
+                                      UnsignedOrNone PackIndex, bool Final) {
+    return getSema().BuildSubstNonTypeTemplateParmExpr(
+        AssociatedDecl, NTTP, Loc, Arg, PackIndex, Final);
+  }
+
 private:
   QualType TransformTypeInObjectScope(TypeLocBuilder &TLB, TypeLoc TL,
                                       QualType ObjectType,
@@ -16459,7 +16468,7 @@ ExprResult TreeTransform<Derived>::TransformSubstNonTypeTemplateParmPackExpr(
 
   TemplateArgument Pack = E->getArgumentPack();
   TemplateArgument Arg = SemaRef.getPackSubstitutedTemplateArgument(Pack);
-  return SemaRef.BuildSubstNonTypeTemplateParmExpr(
+  return getDerived().RebuildSubstNonTypeTemplateParmExpr(
       E->getAssociatedDecl(), E->getParameterPack(),
       E->getParameterPackLocation(), Arg, SemaRef.getPackIndex(Pack),
       E->getFinal());
@@ -16517,10 +16526,10 @@ ExprResult TreeTransform<Derived>::TransformSubstNonTypeTemplateParmExpr(
     Replacement = E->getReplacement();
   }
 
-  return new (SemaRef.Context) SubstNonTypeTemplateParmExpr(
-      Replacement.get()->getType(), Replacement.get()->getValueKind(),
-      E->getNameLoc(), Replacement.get(), AssociatedDecl, E->getIndex(),
-      E->getPackIndex(), E->isReferenceParameter(), E->getFinal());
+  return getDerived().RebuildSubstNonTypeTemplateParmExpr(
+      AssociatedDecl, E->getParameter(), E->getNameLoc(),
+      TemplateArgument(Replacement.get(), /*IsCanonical=*/false),
+      E->getPackIndex(), E->getFinal());
 }
 
 template<typename Derived>

--- a/clang/test/SemaTemplate/concepts.cpp
+++ b/clang/test/SemaTemplate/concepts.cpp
@@ -1658,3 +1658,48 @@ struct {
 void foo() { call(""); }
 
 }
+
+namespace GH170856 {
+
+template <unsigned N, unsigned M> struct symbol_text {
+  consteval symbol_text(const char txt[]) {}
+};
+template <unsigned N> symbol_text(const char (&)[N]) -> symbol_text<1, 1>;
+struct quantity_spec_interface_base {};
+template <symbol_text, auto...> struct named_unit;
+struct quantity_spec_interface : quantity_spec_interface_base {};
+struct : quantity_spec_interface {
+} thermodynamic_temperature;
+
+template <typename T>
+concept QuantitySpec = __is_convertible(T*, quantity_spec_interface_base*);
+template <typename, auto QS>
+concept QuantitySpecOf = QuantitySpec<decltype(QS)>;
+template <typename T, auto QS>
+concept PointOriginFor = QuantitySpecOf<decltype(QS), T::_quantity_spec_>;
+
+template <symbol_text Symbol, auto QS, auto PO>
+struct named_unit<Symbol, QS, PO> {
+  static constexpr auto _point_origin_ = PO;
+};
+template <auto QS> struct absolute_point_origin {
+  static constexpr auto _quantity_spec_ = QS;
+};
+template <auto> struct relative_point_origin {};
+template <class R>
+consteval PointOriginFor<R{}> auto default_point_origin(R) {
+  return R{}._point_origin_;
+}
+template <auto> class quantity_point;
+template <class R> struct point_ {
+  quantity_point<default_point_origin(R{})> operator0();
+};
+template <auto R> point_<decltype(R)> point;
+struct absolute_zero : absolute_point_origin<thermodynamic_temperature> {
+} absolute_zero;
+auto zeroth_kelvin = absolute_zero;
+struct : named_unit<"", thermodynamic_temperature, zeroth_kelvin> {
+} kelvin;
+struct ice_point : relative_point_origin<point<kelvin>> {};
+
+}

